pg_mon.c

```diff
@@ -438,24 +470,104 @@ pgmon_ExecutorEnd(QueryDesc *queryDesc)
 {
     uint64		queryId = queryDesc->plannedstmt->queryId;
 
-        if (queryId != UINT64CONST(0) && queryDesc->totaltime && nesting_level == 0)
+    if (queryId != UINT64CONST(0) && queryDesc->totaltime && nesting_level == 0)
+    {
+            /*
+             * Make sure stats accumulation is done.
+             * (Note: it's okay if several levels of hook all do this.)
+             */
+            InstrEndLoop(queryDesc->totaltime);
+            InstrEndLoop(queryDesc->planstate->instrument);
+
+            /* Save query information */
+            pgmon_exec_store(queryDesc);
+    }
+
+    if (prev_ExecutorEnd)
+            prev_ExecutorEnd(queryDesc);
+    else
+            standard_ExecutorEnd(queryDesc);
+}
+
+/*
+ * ProcessUtility hook
+ */
+#if PG_VERSION_NUM < 130000
+static void pgmon_ProcessUtility(PlannedStmt *pstmt, const char *queryString,
+						   ProcessUtilityContext context, ParamListInfo params,
+						   QueryEnvironment *queryEnv,
+						   DestReceiver *dest, char *completionTag)
+{
+    if (CONFIG_LOG_NEW_QUERY)
+    {
+        switch (nodeTag(pstmt->utilityStmt))
+        {
+            case T_AlterRoleStmt:
+                break;
+            default:
+                ereport(LOG, (errmsg("%s", queryString), errhint("log from pg_mon")));
+                break;
+        }
+    }
```

Comment:
It seems that these lines are copy&pasted 3 times across 3 variants of ProcessUtility hooks for different Postgres versions.

I think it would be better to use the approach like for example in the set_user: https://github.com/pgaudit/set_user/blob/master/compatibility.h#L20

I.e., just define some macros for different versions and use them in the code in order to get rid of duplications.

Reply:
Thanks for the input here!